geode-core/src/main/java/org/apache/geode/internal/cache/TXCommitMessage.java

```diff
@@ -391,9 +416,34 @@ void send(TXLockId lockId) {
                 setRecipientsSendData(Collections.singleton(indivRecip.next()), processor, rcl);
               }
             } else {
-              // Run in normal mode sending to multiple recipients in
-              // one shot
-              setRecipientsSendData(recipients, processor, rcl);
+              Set<InternalDistributedMember> tempNotificationOnlyMembers =
+                  new HashSet<InternalDistributedMember>();
+              Set<InternalDistributedMember> tempTransactionMembers =
+                  new HashSet<InternalDistributedMember>();
+              if (!rcl.isEmpty() && !getNotificationOnlyMembers().isEmpty()) {
+                for (RegionCommit rc : rcl) {
+                  Set<InternalDistributedMember> getNOM = getNotificationOnlyMembers().get(rc);
+                  Set<InternalDistributedMember> getTM = getTransactionMembers().get(rc);
+                  if (getNOM != null && !getNOM.isEmpty()) {
+                    tempNotificationOnlyMembers.addAll(getNOM);
+                  }
+                  if (getTM != null && !getTM.isEmpty()) {
+                    tempTransactionMembers.addAll(getTM);
+                  }
+                }
+                tempNotificationOnlyMembers.removeAll(tempTransactionMembers);
+              }
+
+              if (tempNotificationOnlyMembers.isEmpty()) {
+                // Run in normal mode sending to multiple recipients in one shot
+                setRecipientsSendData(recipients, processor, rcl);
+              } else {
+                recipients.removeAll(tempNotificationOnlyMembers);
```

Comment:
This still has the same issue as multiple TxCommitMessages (for different rcl) being sent to the same remote server. 
Let's say there are only 2 rcls in the tx. First rcl contains 1 RegionCommit on bucket_1. TxCommitMessage (containing the rcl) should be sent to remote server1 as redundant server. (assume server 2 will be the notificationOnlyMember due to cq/subscription etc). This will cause on first rcl being sent to both server1 and server2.

Second rcl also contains 1 RegionCommit on bucket_2. Server 2 is the redundant server, and server 1 needs to be set as notification only member. This will cause the second TxCommitMessage (containing second rcl) being sent to both server1 and server2. 

As a result, some of the RCLs is not being processed - causing data inconsistency. Each server should only receive one TxCommitMessage for a tx.


When processing second rcl (

Reply:
It seems, that this approach to solve this issue on sending side is generating more and more patches. Maybe this shoulb be solved on receiving side as @jvarenina proposed in https://github.com/apache/geode/pull/6477#issuecomment-848723153

Reply:
I have found what was the problem, and pushed fix, but should we also consider other alternative, mentioned above.

Reply:
I think the issue is that the fix was based on the region commits list (for different receivers).
I think you may want to consider to use the information available in tx - for each entry event inside the tx.
FilterRoutingInfo filterRoutingInfo is sent to remote receivers through FarSideEntryOp. It can be queried for the adjunct servers. It is calculated on the primary (tx host node).

At remote receiver site, txCallbackEventFactory.createCallbackEvent is invoked to created the callback event to be used for wan etc. You can use the following information to determine whether to set the tailKey for this event.
txEntryState == null indicates this is a farsider receiver (not host)
eventRegion.isUsedForPartitionedRegionBucket() to determine if a bucket region (before setting as a pr).
use the filterRoutingInfo.getMembers to determine the receive is one of the adjunctServers -- to determine if tailKey should be reset.

Reply:
Hi @pivotal-eshu ,

Thank you for the help on this issue. We really appreciate it.
I have create the draft PR with the solution that it based on what you proposed in above comment: https://github.com/apache/geode/pull/6646

Could you please check it and comment?

Just to share with you why I haven't used some of the proposed checks in the solution:

`internalRegion.getCacheDistributionAdvisor().adviseTX() `-> I haven't used this one because adviseTX only works with BucketRegion. Since bucket region is not available on adjunct server it not possible to get primary and redundant servers with this function there.

`txEntryState == null` -> Not sure that this is needed because impacted part of code is executed only on farside.

`internalRegion.isUsedForPartitionedRegionBucket() `-> I haven't used this check because it returns "false" on adjunct member. I think that this is already checked with `this.internalRegion instanceof PartitionedRegion` few lines before the impact. When server hosts bucket regions then `internalRegion` will be of type `BucketRegion`.